File: src/main/java/xyz/oribuin/playerwarps/command/argument/WarpArgumentHandler.java
package xyz.oribuin.playerwarps.command.argument;

import dev.rosewood.rosegarden.command.framework.Argument;
import dev.rosewood.rosegarden.command.framework.ArgumentHandler;
import dev.rosewood.rosegarden.command.framework.CommandContext;
import dev.rosewood.rosegarden.command.framework.InputIterator;
import org.bukkit.entity.Player;
import xyz.oribuin.playerwarps.PlayerWarpsPlugin;
import xyz.oribuin.playerwarps.manager.DataManager;
import xyz.oribuin.playerwarps.model.Warp;

import java.util.List;

public class WarpArgumentHandler extends ArgumentHandler<Warp> {

    public WarpArgumentHandler() {
        super(Warp.class);
    }

    @Override
    public Warp handle(CommandContext context, Argument argument, InputIterator inputIterator) throws HandledArgumentException {
        String input = inputIterator.next();

        Warp warp = PlayerWarpsPlugin.get().getManager(DataManager.class)
                .getWarps()
                .get(input);

        if (warp == null)
            throw new HandledArgumentException("argument-handler-warp");

        return warp;
    }

    @Override
    public List<String> suggest(CommandContext context, Argument argument, String[] args) {
        DataManager manager = PlayerWarpsPlugin.get().getManager(DataManager.class);

        if (context.getSender() instanceof Player player && !player.hasPermission("playerwarps.bypass")) {
            return manager.getOwned(player.getUniqueId())
                    .stream()
                    .map(Warp::getId)
                    .toList();
        }

        return manager.getWarps().keySet().stream().toList();
    }

}

File: src/main/kotlin/xyz/oribuin/playerwarps/command/argument/WarpArgumentHandler.kt
package xyz.oribuin.playerwarps.command.argument

import dev.rosewood.rosegarden.RosePlugin
import dev.rosewood.rosegarden.command.framework.ArgumentParser
import dev.rosewood.rosegarden.command.framework.RoseCommandArgumentHandler
import dev.rosewood.rosegarden.command.framework.RoseCommandArgumentInfo
import dev.rosewood.rosegarden.utils.StringPlaceholders
import xyz.oribuin.playerwarps.manager.WarpManager
import xyz.oribuin.playerwarps.util.WarpUtils.getManager
import xyz.oribuin.playerwarps.warp.Warp

class WarpArgumentHandler(rosePlugin: RosePlugin) : RoseCommandArgumentHandler<Warp>(rosePlugin, Warp::class.java) {

    @Throws(HandledArgumentException::class)
    override fun handleInternal(argumentInfo: RoseCommandArgumentInfo, argumentParser: ArgumentParser): Warp {
        val input = argumentParser.next()

        return rosePlugin.getManager<WarpManager>()
            .getWarp(input.lowercase())
            ?: throw HandledArgumentException("argument-handler-warps", StringPlaceholders.single("input", input))
    }

    override fun suggestInternal(argumentInfo: RoseCommandArgumentInfo, argumentParser: ArgumentParser): List<String> {
        argumentParser.next()

        return this.rosePlugin.getManager<WarpManager>()
            .getWarps()
            .map { it.name.lowercase() } // Get the name of each warp
            .ifEmpty { listOf("<no loaded warps>") } // If there are no warps, return a placeholder
    }
}
